### PR TITLE
Checkpointing on structure factor

### PR DESCRIPTION
**Pull request by rousseab**
I refactored how the Kolmogorov Smirnov (KS) values are computed. This is now done in the model lightning module instead of in a callback. This is necessary in order to log these values and use them for checkpoint monitoring and/or early stopping (the call order of callbacks / model hooks is indertermined, so we can't rely on a callback to log metrics). We now get more telemetry during training on just what the heck is going on (Kudos on the structure factor, btw, it's *very* nice).

Screenshot 2024-09-23 at 8 03 45 AM

What used to be the sampling callback has been refactored to visualize and write to disk what is already calculated in the model lightning module. This relies on internal properties (ie, "states"), which is not super clean, but it's the PL way.  

I also refactor the configuration file in a way that breaks backward compatibility. I didn't not fix every example / sanity check / off-broadway code bits we have. See the following file for an example:
examples/config_files/diffusion/config_diffusion_mlp.yaml

Future possible improvements might be to compute the validation set distances and energies *only once* (write them in a tmp cache file perhaps) instead of wasting time redoing this at every epoch.

### CODE REVIEW

File: crystal_diffusion/models/position_diffusion_lightning_model.py [hunk 1]
```diff
@@ -217,15 +267,17 @@ def _get_target_normalized_score(
...
+            noisy_relative_coordinates - real_relative_coordinates
+        )
         target_normalized_scores = get_sigma_normalized_score(
             delta_relative_coordinates, sigmas, kmax=self.hyper_params.kmax_target_score
         )
         return target_normalized_scores
 
     def training_step(self, batch, batch_idx):
         """Runs a prediction step for training, returning the loss."""
+        logger.info(f"  - Starting training step with batch index {batch_idx}")
```
Comment (sblackburn86) [inline crystal_diffusion/models/position_diffusion_lightning_model.py:280]: as discussed - probably better to remove

File: crystal_diffusion/models/position_diffusion_lightning_model.py [hunk 2]
```diff
@@ -235,18 +287,59 @@ def training_step(self, batch, batch_idx):
...
+            on_step=False,
+            on_epoch=True,
+            prog_bar=True,
+        )
+
+        if not self.draw_samples:
+            return output
+
+        if self.metrics_parameters.compute_energies:
+            logger.info("        * registering reference energies")
```
Comment (sblackburn86) [inline crystal_diffusion/models/position_diffusion_lightning_model.py:321]: same as for training step logger.info

File: crystal_diffusion/models/position_diffusion_lightning_model.py [hunk 3]
```diff
@@ -235,18 +287,59 @@ def training_step(self, batch, batch_idx):
...
+            reference_energies = batch["potential_energy"]
+            self.energy_ks_metric.register_reference_samples(reference_energies.cpu())
+
+        if self.metrics_parameters.compute_structure_factor:
+            logger.info("        * registering reference distances")
```
Comment (sblackburn86) [inline crystal_diffusion/models/position_diffusion_lightning_model.py:326]: remove,debug or optional (i.e. remove)

File: crystal_diffusion/models/position_diffusion_lightning_model.py [hunk 4]
```diff
@@ -235,18 +287,59 @@ def training_step(self, batch, batch_idx):
...
+            reference_distances = compute_distances_in_batch(
+                cartesian_positions=cartesian_positions,
+                unit_cell=basis_vectors,
+                max_distance=self.metrics_parameters.structure_factor_max_distance,
+            )
+            self.structure_ks_metric.register_reference_samples(
+                reference_distances.cpu()
+            )
+
+        logger.info(f"         Done validation step with batch index {batch_idx}")
```
Comment (sblackburn86) [inline crystal_diffusion/models/position_diffusion_lightning_model.py:342]: same
